portfolio-project-matching-app/backend/dao.js

```diff
@@ -1,246 +1,351 @@
+        return -1;
+    }
+    // handle case where id1 does not exist in coll1
+    else if (!id1Snap.exists()) {
+        console.log(`Invalid '${coll1}' id: '${id1}' does not exist`);
+        return -1;
+    }
+    // handle case of projects_users where project is not open
+    else if (coll === 'projects_users' && !id1Snap.data().open) {
+        console.log(`Invalid '${coll1}' id: ${id1} is at capacity`);
+        return -1;
+    }
+    // handle case where id2 does not exist in coll2
+    else if (!id2Snap.exists()) {
+        console.log(`Invalid '${coll2}' id: '${id2}' does not exist`);
+        return -1;
+    }
+    // handle case where id1_id2 already exists in coll
+    else if (collSnap.exists()) {
+        console.log(`Invalid id1 id2 combo: '${id1}_${id2}' already exists in '${coll}'`);
+        return -1;
+    }
+    // handle case inputs are valid 
+    else {
+        // build association document to send to Firebase
+        const payload = getPayload(coll, id1, id2);
+        const newDocSnap = await addNewDocWithId(coll, `${id1}_${id2}`, payload);
+        console.log(`Created '${coll}' document with id: ${newDocSnap.id}`);
+        // handle case of projects_users
+        if (coll === 'projects_users') {
+            // create update payload for incremented census
+            const payload = {
+                census: id1Snap.data().census + 1,
+            }
+            // handle case where project needs closed
+            const payloadNeedsUpdated = id1Snap.data().census === id1Snap.data().capacity - 1;
+            if (payloadNeedsUpdated){
+                payload.open = false;
+            }
+            const projectSnapNew = await updateDoc(coll1, id1, payload);
+            console.log(`Updated '${coll1}' document id '${id1}' census to ${projectSnapNew.data().census}`);
+            if (payloadNeedsUpdated) {
+                console.log(`Closed '${coll1}' document id '${id1}'`);
+            }
+        }
+        return newDocSnap;
     }
-    return technologies;
-}
-
-const getUsersByProjectId = async (projectId) => {
-    /*
-    DESCRIPTION:    retrieves users associated with specified project ID
-
-    INPUT:          desired project document ID in string format
-
-    RETURN:         array of User objects associated with project
-    */
-    // get filtered collection snapshot
-    const projectsUsersSnap = await getCollectionSnapshotByCriteria('projects_users', 'project_id', '==', projectId);
-    
-    // loop through associated documents, adding User objects to array
-    const users = [];
-    for (const doc of projectsUsersSnap.docs) {
-        const userRef = await getDocSnapshotById('users', doc.data().user_id);
-        const user = new User(userRef);
-        users.push(user);
-    }
-    return users;
 }
 
-/*
-    USER LEVEL QUERIES
-*/
+// helper functions, don't export
 
-const getAllUsers = async () => {
+const getPayload = (coll, id1, id2) => {
```

Comment:
Little helper function to keep things cleaner in the main function. I could do a ton more work leveraging helpers in tidying things up a bit.